---
title: "Comparing Samples"
author: "Luigi Marchionni"
date: "9/25/2021"
output: html_document
---

# Preliminary operations

```{r compileRMD, echo=FALSE, eval=TRUE}
### Libraries necessary for knitting
library(rmarkdown)
```

## Load the required general libraries for typesetting

```{r setup}
### Clean the workspace
rm(list=ls())
ls()
### Load the required libraries
library(knitr, quietly = TRUE)
knitr::opts_chunk$set(echo = TRUE)
### To run and output elsewhere
knitPath <- "/home/idies/workspace/practical_genomics/day3/"
# knitPath <- "."
print(knitPath)
opts_knit$set(root.dir = knitPath) #tidy=TRUE, tidy.opts=list(width.cutoff=180))
```

## Set the correct working directory

```{r}
### Finding your current working directory
getwd()
### Setting the working directory (change this)
setwd(knitPath)
### Check working directory
getwd()
```

# Comparing multiple samples
Comparing multiple samples, and basically making sure to some extent that your observations generalize when n > 1 is possibly one of the most important question one might want to ask. With single cell technologies, this translates to be able to ask questions like "are genes differentially expressed, or cell populations differentially represented between phenotypes?". These differential analyses can be cathegorized into two distinct analysis types: differential expression (DE) and differential abundance (DA).

## Preparing the data
To exempify these analyses, we will analyze samples from early mouse embryo development (day E8.5), comparing chimeric embryos generated by injecting td-Tomato-positive embryonic stem cells (ESCs) to their wild-type (WT) blastocyst countepart. The aim of this study is to identify the differences in lineage commitment determined by the injection procedure itself (see, Pijuan-Sala et al. Nature, 2019). In this experiment, a paired design was used, with three replicates of the two matched samples, generating scRNA-seq data using the 10X Genomics protocol.

We will start by loading and pre-processing the data.

```{r, eval=FALSE, echo=TRUE}
### Loading data from the Bioc packages
library(MouseGastrulationData, quietly = TRUE)
### Get the data of interest
sce.chimera <- WTChimeraData(samples=5:10)
### Show the data of interest
sce.chimera
### Annotate the genes
library(scater, quietly = TRUE)
rownames(sce.chimera) <- uniquifyFeatureNames(rowData(sce.chimera)$ENSEMBL, rowData(sce.chimera)$SYMBOL)
### Perform some quality control: dropping unwanted features
drop <- sce.chimera$celltype.mapped %in% c("stripped", "Doublet")
table(drop)
sce.chimera <- sce.chimera[,!drop]
### Perform normalization
sce.chimera <- logNormCounts(sce.chimera)
### Model variance
library(scran, quietly = TRUE)
dec.chimera <- modelGeneVar(sce.chimera, block=sce.chimera$sample)
chosen.hvgs <- dec.chimera$bio > 0
### Now let's merge the distinct datasets and check correcto for batches
library(batchelor, quietly = TRUE)
### Set seed
set.seed(01001001)
### Correct
merged <- correctExperiments(
  sce.chimera, batch=sce.chimera$sample, subset.row=chosen.hvgs,
  PARAM=FastMnnParam(
    merge.order=list(
      list(1,3,5), # WT (3 replicates)
      list(2,4,6)  # td-Tomato (3 replicates)
      )
    )
  )
### Perform clustering
g <- buildSNNGraph(merged, use.dimred="corrected")
clusters <- igraph::cluster_louvain(g)
colLabels(merged) <- factor(clusters$membership)
### Perform dimensionality reduction
merged <- runTSNE(merged, dimred="corrected", external_neighbors=TRUE)
merged <- runUMAP(merged, dimred="corrected", external_neighbors=TRUE)
### Display merged data and save
merged
### Save it
save(merged, file = "./chimera.sce.merged.rda")
```

# Session information
```{r}
sessionInfo()
```






